### ISSUE TITLE
Fl_Simple_Terminal::append_ansi() not bounded on length

### ISSUE THREAD

**Issue by Prof9** [fltk/fltk#854]
**Describe the bug**
`Fl_Simple_Terminal::append_ansi()` accepts a string `len` parameter, however it is not bounded on it. Instead the function will keep appending until a null byte is encountered. This is a problem because the function allocates some temp buffers and does use the `len` parameter there. Therefore if the string is not null-terminated, `append_ansi()` will write past the end of the allocated buffer.

Offending line is here:
https://github.com/fltk/fltk/blob/89454508a2024da5fde5db309ede45a0af9cdeb3/src/Fl_Simple_Terminal.cxx#L806

**To Reproduce**
I'm using FLTK through Rust bindings, so forgive me for using a Rust example. The actual problem is in FLTK proper, however.
```rust
let mut term = fltk::text::SimpleTerminal::default_fill();
let foobar: &str = "foobar";
let foo: &[u8] = unsafe { std::slice::from_raw_parts(foobar.as_ptr(), 3) };
term.append2(foo);  // Internally calls Fl_Simple_Terminal::append(foo.as_ptr(), foo.len())
```
This can result in one of two things:
* A segfault
* `foobar` is printed to the terminal along with anything else that happens to be after it in memory.

**Expected behavior**
Only `foo` is printed.

**FLTK Version**
FLTK 1.4.0 (specifically Git 01d30ed9cc4c080dea45ee85ec7af76cbd7e61bb) used via fltk-rs 1.4.17.

**Comment by MoAlyousef** [comment]
Hi @Albrecht-S 

@jgriffitts has recently added Fl_Terminal to fltk-rs. I had been planning to alias SimpleTerminal to point to Terminal once FLTK removes Fl_Simple_Terminal, then remove SimpleTerminal when FLTK 1.4 is released (as I will start working on fltk-rs 2.0).

**Comment by Albrecht-S** [comment]
> I had been planning to alias SimpleTerminal to point to Terminal once FLTK removes Fl_Simple_Terminal, then remove SimpleTerminal when FLTK 1.4 is released (as I will start working on fltk-rs 2.0).

@MoAlyousef This looks like a catch-22 issue. We left `Fl_Simple_Terminal` for now in FLTK 1.4 (Git master) only for others to be able to migrate to `Fl_Terminal` as soon as possible w/o having to do it **suddenly** w/o a warning.

I suggest to remove `(Fl_)Simple_Terminal` from `fltk-rs` as soon as you can find the time to do it (i.e. from `cfltk`, if I understand correctly).

Note that we're striving to release 1.4 by the end of this year. However `Fl_Simple_Terminal` will be removed earlier, maybe in one or two weeks, but definitely some time before 1.4.0 will be released.

**Comment by Prof9** [comment]
Just to clarify, I was not using `Fl_Simple_Terminal::append_ansi()` directly. Rather, I was using `Fl_Simple_Terminal::append()`, which calls `Fl_Simple_Terminal::append_ansi()` internally.
https://github.com/fltk/fltk/blob/f72748bb4563c07322846ae92cc9c02c0b448e5b/src/Fl_Simple_Terminal.cxx#L907-L911

**Comment by jgriffitts** [comment]
For what it's worth: I was working on the Rust bindings for Fl_Terminal last week, and I tested for this specific problem on all the append() methods. I was looking for bugs in the Rust wrapper, not in fltk itself, but everything was OK at that level.

**Comment by Albrecht-S** [comment]
@jgriffitts Thank you for your comment (and testing/implementing `Fl_Terminal` in `Rust`). Your confirmation is very much appreciated.

**Comment by erco77** [comment]
Ya, as far as I know nothing mentioned here applies to Fl_Terminal.. it's a very different thing internally.

Indeed it appears Fl_Simple_Terminal's append_ansi() doesn't count the 'len' parameter down during the while() loop that walks the user's string; an omission obviously. I'll apply a fix (even though the widget will be removed), just in case the widget ever needs to take on a life of its own in user code for some reason. Will follow up.

**Comment by erco77** [comment]
OK, fix applied in c3849b0 so append_ansi() (and friends) respect the len argument, and added some docs about how len can be -1 to imply an internal strlen() calculation.
